Adding documentation on device docs (#6325)
PennyLaneAI/pennylane: doc/development/plugins.rst

different numbers of shots to use when calculating the final expecation value.

>>> tape0 = qml.tape.QuantumScript([], [qml.expval(qml.PauliX(0))], shots=(5, 500, 1000))
>>> tape0.shots.shot_vector
(ShotCopies(5 shots x 1),
 ShotCopies(500 shots x 1),
 ShotCopies(1000 shots x 1))
>>> list(tape0.shots)
[5, 500, 1000]
>>> list(tape0.shots.bins())
[(0, 5), (5, 505), (505, 1505)]
>>> dev.execute(tape0)
(0.2, -0.052, -0.014)

The first number ``0.2`` is calculated with 5 shots, the second ``-0.052`` is calculated with 500 shots, and
``-0.014`` is calculated with 1000 shots.  All 1,505 shots can be requested together in one batch, but the post
processing into the expecation value is done with shots ``0:5``, ``5:505``, and ``505:1505`` respectively.

Wires
-----

Devices can now either:

1) Strictly use wires provided by the user on initialization ``device(name, wires=wires)``
2) Infer the number and ordering of wires provided by the submitted circuit.
3) Strictly require specific wire labels

Option 2 allows workflows to change the number and labeling of wires over time, but sometimes users want

> to enfore a wire convention and labels. If a user does provide wires, the :meth:`~.devices.Device.preprocess` should

Reviewer: This might be confusing before explaining what `preprocess` does

Author: moving the description of wires below preprocessing.

validate that submitted circuits only have wires in the requested range.

>>> dev = qml.device('default.qubit', wires=1)
>>> circuit = qml.tape.QuantumScript([qml.CNOT((0,1))], [qml.state()])
>>> dev.preprocess()[0]((circuit, ))
WireError: Cannot run circuit(s) of default.qubit as they contain wires not found on the device.

> PennyLane wires can be any hashable object, where wire labels are distinguished by their equality and hash.

Reviewer: I assume this corresponds to option 2 mentioned above? Can we include an example here?

Author: wire hashability?  That corresponds to any option.

Reviewer: Huh I don't know why it was added to this line but I meant to comment on the `map_to_standard_wires` thing below.

If working with successive integers (``0``, ``1``, ``2``, ...) is preferred internally,
the :meth:`~.QuantumScript.map_to_standard_wires` method can be used inside of 
the :meth:`~.devices.Device.execute` method.  The :class:`~.map_wires` transform can also 
map the wires of the submitted circuit to internal labels.

Sometimes hardware qubit labels cannot be arbitrarily mapped without a changing in behaviour.
Connectivity, noise, performance, and
other constraints can make it so that operations on qubit one cannot be arbitrarily exchanged with the same operation
of qubit two.  In such a situation, the device can hard code a list of the only acceptable wire labels. In such a case, it
will be on the user to deliberately map wires if they wish such a thing to occur.

>>> qml.device('my_hardware').wires
<Wires = [0, 1, 2, 3]>
>>> qml.device('my_hardware', wires=(10, 11, 12, 13))
DeviceError: Device my_hardware cannot internally map wires, as the labels 0, 1, 2, 3 indicate unique qubits

Preprocessing
-------------

> The preprocessing method has two main responsibilities:

Reviewer: Maybe to put it into perspective, we can briefly explain that a PL circuit is first transformed by the transform program returned by `preprocess`, and then the transformed circuits are sent to `execute` to be executed. This will help the developer understand the purpose of `preprocess` better.

These two tasks can be extracted into private methods or helper functions if that improves source code organization.

See the section on the **Execution Config** below for more information on step 2.

Devices may define their own transforms following the description in the ``transforms`` module,
or can include in-built transforms such as:

.. currentmodule:: pennylane.devices.preprocess
.. autosummary::
    :toctree: api

    decompose
    validate_observables
    validate_measurements
    validate_device_wires
    validate_multiprocessing_workers
    validate_adjoint_trainable_params
    no_sampling

.. currentmodule:: pennylane
.. autosummary::
    :toctree: api

    defer_measurements
    dynamic_one_shot
    transforms.broadcast_expand
    transforms.split_non_commuting
    transforms.transpile
    transforms.diagonlize_measurements
    transforms.split_to_single_terms

Once a program is created, an individual transform can be added to the program with the
:meth:`~.TransformProgram.add_transform` method.

.. code-block:: python

    from pennylane.devices.preprocess import validate_device_wires

>     program = qml.transforms.core.TransformProgram()

Author: fleshing out this example to include all the high priority transforms.

Preprocessing and validation can also exist inside the :meth:`~devices.Device.execute` method, but placing them
in the preprocessing program has several benefits. Validation can happen earlier, leading to fewer resources
spent before the error is raised.  Users can inspect, draw, and spec out the tapes at different stages throughout
preprocessing. This provides users a better awareness of what the device is actually executing. When device
gradients are used, the preprocessing transforms are tracked by the machine learning interfaces.  With the
ML framework tracking the classical component of preprocessing, the device does not need to manually track the
classical component of any decompositions or compilation. 

Even with these benefits, devices can still opt to
place some transforms inside the ``execute`` method.  For example, ``default.qubit`` maps wires to simulation indices
inside ``execute`` instead of in ``preprocess``.

The :meth:`~.devices.Device.execute` can assume that device preprocessing has been run on the input
tapes, and has no obligation to re-validate the input or provide sensible error messages.

> >>> op = qml.Permute([2,1,0], wires=(0,1,2))

Reviewer: It is not so clear what this example is trying to demonstrate.

``use_device_jacobian_product`` indicates a request to call ``compute_vjp`` instead of ``compute_derivatives``.  Note that 
if ``use_device_jacobian_product`` is ``True``, this takes precedence over calculating the full jacobian.

>>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
>>> processed_config = qml.device('default.qubit').preprocess(config)[1]
>>> processed_config.use_device_jacobian_product
True
>>> processed_config.use_device_gradient
True
>>> processed_config.grad_on_execution
True

Execution
---------

For documentation on the expected result type output, please refer to :ref:`ReturnTypeSpec`.

The device API allows individual devices to calculate results in whatever way makes sense for
the individual device. With this freedom over the implementation does come more responsibility
to handle each stage in the process. 

PennyLane does provide some helper functions to assist in executing
circuits. Any ``StateMeasurement`` has ``process_state`` and ``process_density_matrix`` methods for
classical post-processsing of a state vector or density matrix, and ``SampleMeasurement``'s implement

> both ``process_samples`` and ``process_counts``.  The ``pennylane.devices.qubit`` module also contains

Reviewer: Can we also add examples to explain how this fits in the implementation of the `execute` method?

Author: better?

.. role:: html(raw)
   :format: html

Building a plugin
=================

> For adding a plugin that inherits from the legacy interface, see :doc:`/development/legacy_plugins`.

Author: suggestion is renaming the legacy file.

.. _plugin_overview:

.. important::

    In your plugin module, **standard NumPy** (*not* the wrapped Autograd version of NumPy)
    should be imported in all places (i.e., ``import numpy as np``).

> Writing your own PennyLane plugin, to allow an external quantum library to take advantage of the
> automatic differentiation ability of PennyLane, is a simple and easy process. In this section, we discuss

Reviewer 2: ```suggestion
PennyLane plugins allow an external quantum library to take advantage of the automatic differentiation ability of PennyLane. Writing your own plugin is a simple and easy process. In this section, we discuss
```

minimal device, we recommend looking at the implementation in ``pennylane/devices/reference_qubit.py``.

Creating your device
--------------------

In order to define a custom device, you only need to override the :meth:`~.devices.Device.execute` method.

.. code-block:: python

    from pennylane.devices import Device, DefaultExecutionConfig
    from pennylane.tape import QuantumScript, QuantumScriptOrBatch

    class MyDevice(Device):
        """My Documentation."""

        def execute(self, circuits: QuantumScriptOrBatch, execution_config: "ExecutionConfig" = DefaultExecutionConfig):
            # your implementation here.

For example:

.. code-block:: python

    class MyDevice(Device):
        """My Documentation."""

        def execute(self, circuits: QuantumScriptOrBatch, execution_config: "ExecutionConfig" = DefaultExecutionConfig):
            return 0.0 if isinstance(circuits, qml.tape.QuantumScript) else tuple(0.0 for c in circuits)

    dev = MyDevice()

    @qml.qnode(dev)
    def circuit():
        return qml.state()

    circuit()

This execute method works in tandem with the optional ``Device.preprocss``, described below in more detail.

> Preprocessing turns generic circuits into ones supported by the device or errors out. Execution then

Reviewer: ```suggestion
Preprocessing turns generic circuits into ones supported by the device, or raises an error if the circuit is invalid. Execution then
```

turns those supported circuits into numerical results. 

> In a more minimal example, for any initial batch and config, we expect to be able to do:

Reviewer: ```suggestion
In a more minimal example, for any initial batch of quantum tapes and a config object, we expect to be able to do:
```

.. code-block:: python

    transform_program, execution_config = dev.preprocess(initial_config)
    circuit_batch, postprocessing = transform_program(initial_circuit_batch)
    results = dev.execute(circuit_batch, execution_config)
    final_results = postprocessing(results)

> Shots

Reviewer: It just came to me. We might add a comment here about how `shots.total_shots` being `None` corresponds to analytic execution, and that the device might want to raise an error if it is not supported.

>     fn(dev.execute(batch, execution_config))

Reviewer: What is `fn` here?

Author: renaming to `postprocessing` and separately applying it into a different step.

>     transform_program, execution_config = dev.preprocess(initial_config)
>     batch, fn = transform_program(initial_batch)
>     fn(dev.execute(batch, execution_config))

Reviewer: ```suggestion
    transform_program, execution_config = dev.preprocess(initial_config)
    circuit_batch, fn = transform_program(initial_circuit_batch)
    fn(dev.execute(circuit_batch, execution_config))
```

>>> tape1 = qml.tape.QuantumScript([], [qml.sample(wires=0)], shots=10)
>>> dev = qml.device('default.qubit')
>>> dev.execute((tape0, tape1))
(array([0, 0, 0, 0, 0]), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))

The :class:`~.measurements.Shots` class describes the shots. Users can optionally specify a shot vector, or
different numbers of shots to use when calculating the final expecation value.

>>> tape0 = qml.tape.QuantumScript([], [qml.expval(qml.PauliX(0))], shots=(5, 500, 1000))
>>> tape0.shots.shot_vector
(ShotCopies(5 shots x 1),
 ShotCopies(500 shots x 1),
 ShotCopies(1000 shots x 1))
>>> list(tape0.shots)
[5, 500, 1000]
>>> list(tape0.shots.bins())
[(0, 5), (5, 505), (505, 1505)]
>>> dev.execute(tape0)
(0.2, -0.052, -0.014)

The first number ``0.2`` is calculated with 5 shots, the second ``-0.052`` is calculated with 500 shots, and
``-0.014`` is calculated with 1000 shots.  All 1,505 shots can be requested together in one batch, but the post
processing into the expecation value is done with shots ``0:5``, ``5:505``, and ``505:1505`` respectively.

Preprocessing
-------------

The preprocessing method has two main responsibilities:

> 1) Create a :class:`~.TransformProgram` capable of turning an arbitrary :class:`~.QuantumScript` into one supported the ``execute`` method.

Reviewer: ```suggestion
1) Create a :class:`~.TransformProgram` capable of turning an arbitrary batch of :class:`~.QuantumScript` into a new batch of tapes supported the ``execute`` method.
```

Once the transform program has been applied to a batch of circuits, that batch should be run via ``Device.execute`` without error.

.. code-block:: python

    transform_program, execution_config = dev.preprocess(initial_config)
    batch, fn = transform_program(initial_batch)
    fn(dev.execute(batch, execution_config))

These two tasks can be extracted into private methods or helper functions if that improves source code organization.

See the section on the **Execution Config** below for more information on step 2.

Once a program is created, an individual transform can be added to the program with the
:meth:`~.TransformProgram.add_transform` method.

.. code-block:: python

    from pennylane.devices.preprocess import validate_device_wires, validate_measurements, decompose

    program = qml.transforms.core.TransformProgram()
    program.add_transform(validate_device_wires, wires=qml.wires.Wires((0,1,2)), name="my_device")
    program.add_transform(validate_measurements, name="my_device")
    program.add_transform(qml.defer_measurements)
    program.add_transform(qml.transforms.split_non_commuting)

>     def supports_operation(op): return getattr(op, "name", None) in operation_names

Reviewer: ```suggestion
    def supports_operation(op): 
        return getattr(op, "name", None) in operation_names
        
```

    def supports_operation(op): 
        return getattr(op, "name", None) in operation_names
        
    program.add_transform(decompose, stopping_condition=supports_operation, name="my_device")
    program.add_transform(qml.transforms.broadcast_expand)

Preprocessing and validation can also exist inside the :meth:`~devices.Device.execute` method, but placing them
in the preprocessing program has several benefits. Validation can happen earlier, leading to fewer resources

> spent before the error is raised.  Users can inspect, draw, and spec out the tapes at different stages throughout

Reviewer 2: ```suggestion
spent before the error is raised. Users can inspect, draw, and spec out the tapes at different stages throughout
```

preprocessing. This provides users a better awareness of what the device is actually executing. When device

> gradients are used, the preprocessing transforms are tracked by the machine learning interfaces.  With the

Reviewer 2: ```suggestion
gradients are used, the preprocessing transforms are tracked by the machine learning interfaces. With the
```

ML framework tracking the classical component of preprocessing, the device does not need to manually track the
classical component of any decompositions or compilation. For example,

>>> @qml.qnode(qml.device('reference.qubit', wires=2))
... def circuit(x):
...     qml.IsingXX(x, wires=(0,1))
...     qml.CH((0,1))
...     return qml.expval(qml.X(0))
>>> print(qml.draw(circuit, level="device")(0.5))
0: ─╭●──RX(0.50)─╭●────────────╭●──RY(-1.57)─┤  <Z>
1: ─╰X───────────╰X──RY(-0.79)─╰Z──RY(0.79)──┤     

Allows the user to see that both ``IsingXX`` and ``CH`` are decomposed by the device, and that
the diagonalizing gates for ``qml.expval(qml.X(0))`` are applied.

Even with these benefits, devices can still opt to

> place some transforms inside the ``execute`` method.  For example, ``default.qubit`` maps wires to simulation indices

Reviewer 2: ```suggestion
place some transforms inside the ``execute`` method. For example, ``default.qubit`` maps wires to simulation indices
```

inside ``execute`` instead of in ``preprocess``.

The :meth:`~.devices.Device.execute` can assume that device preprocessing has been run on the input
tapes, and has no obligation to re-validate the input or provide sensible error messages. In the below example,
we see `default.qubit` erroring out when unsupported operations and unsupported measurements are 

>>> op = qml.Permute([2,1,0], wires=(0,1,2))
>>> tape = qml.tape.QuantumScript([op], [qml.probs(wires=(0,1))])
>>> qml.device('default.qubit').execute(tape)
MatrixUndefinedError:
>>> tape = qml.tape.QuantumScript([], [qml.density_matrix(wires=0)], shots=50)
>>> qml.device('default.qubit').execute(tape)
AttributeError: 'DensityMatrixMP' object has no attribute 'process_samples'

Devices may define their own transforms following the description in the ``transforms`` module,
or can include in-built transforms such as:

.. currentmodule:: pennylane.devices.preprocess
    :toctree: api

    decompose
    validate_observables
    validate_measurements
    validate_device_wires
    validate_multiprocessing_workers
    validate_adjoint_trainable_params
    no_sampling

> .. currentmodule:: pennylane

Author: ```suggestion
.. currentmodule:: pennylane
.. toctree::
```

> .. toctree::

Reviewer: ```suggestion
.. autosummary::
    :toctree: api
```
I think this was correct afterall?

Author: that was what the original version was, and that caused warnings too.

    transforms.transpile
    transforms.diagonalize_measurements
    transforms.split_to_single_terms

Wires
-----

Devices can now either:

1) Strictly use wires provided by the user on initialization ``device(name, wires=wires)``
2) Infer the number and ordering of wires provided by the submitted circuit.
3) Strictly require specific wire labels

Option 2 allows workflows to change the number and labeling of wires over time, but sometimes users want
to enfore a wire convention and labels. If a user does provide wires, the :meth:`~.devices.Device.preprocess` should
validate that submitted circuits only have wires in the requested range.

>>> dev = qml.device('default.qubit', wires=1)
>>> circuit = qml.tape.QuantumScript([qml.CNOT((0,1))], [qml.state()])
>>> dev.preprocess()[0]((circuit, ))
WireError: Cannot run circuit(s) of default.qubit as they contain wires not found on the device.

PennyLane wires can be any hashable object, where wire labels are distinguished by their equality and hash.
If working with successive integers (``0``, ``1``, ``2``, ...) is preferred internally,
the :meth:`~.QuantumScript.map_to_standard_wires` method can be used inside of 

> the :meth:`~.devices.Device.execute` method.  The :class:`~.map_wires` transform can also 

Reviewer 2: ```suggestion
the :meth:`~.devices.Device.execute` method. The :class:`~.map_wires` transform can also 
```

Author: sorry... a little to hard to change such an ingrained habit...

map the wires of the submitted circuit to internal labels.

Sometimes hardware qubit labels cannot be arbitrarily mapped without a changing in behaviour.
Connectivity, noise, performance, and
other constraints can make it so that operations on qubit one cannot be arbitrarily exchanged with the same operation

> of qubit two.  In such a situation, the device can hard code a list of the only acceptable wire labels. In such a case, it

Reviewer 2: ```suggestion
of qubit two. In such a situation, the device can hard code a list of the only acceptable wire labels. In such a case, it
```

<Wires = [0, 1, 2, 3]>
>>> qml.device('my_hardware', wires=(10, 11, 12, 13))
DeviceError: Device my_hardware cannot internally map wires, as the labels 0, 1, 2, 3 indicate unique qubits

> To implement such validation, a device developer can include a custom transform in their preprocessing:

Reviewer: Why not simply use the `validate_device_wires` provided by PL?

Author: different error message. 

Though actually such an error message should be handled on initialization.  Let me fix that.

While not yet integrated at the top user level, we aim to allow dynamic configuration of the device.

>>> dev = qml.device('default.qubit')
>>> config = qml.devices.ExecutionConfig(device_options={"rng": 42})
>>> tape = qml.tape.QuantumTape([qml.Hadamard(0)], [qml.sample(wires=0)], shots=10)
>>> dev.execute(tape, config)
array([1, 0, 1, 1, 0, 1, 1, 1, 0, 0])
>>> dev.execute(tape, config)
array([1, 0, 1, 1, 0, 1, 1, 1, 0, 0])

By pulling options from this dictionary instead of from device properties,
we unlock two key pieces of functionality.

1) Track and specify the exact configuration of the execution by only inspecting the ``ExecutionConfig`` object
2) Dynamically configure the device over the course of a workflow.

**Workflow Configuration:**

Note that these properties are only applicable to devices that provided derivatives or VJPs. If your device
does not provide derivatives, you can safely ignore these properties.

The workflow options are ``use_device_gradient``, ``use_device_jacobian_product``, and ``grad_on_execution``. 
``use_device_gradient=True`` indicates that workflow should request derivatives from the device. 
``grad_on_execution=True`` indicates a preference

> to use ``execute_and_compute_derivatives`` instead of ``execute`` followed by ``compute_derivatives``.  And finally

Reviewer 2: ```suggestion
to use ``execute_and_compute_derivatives`` instead of ``execute`` followed by ``compute_derivatives``. Finally,
```

> ``use_device_jacobian_product`` indicates a request to call ``compute_vjp`` instead of ``compute_derivatives``.  Note that 

Reviewer 2: ```suggestion
``use_device_jacobian_product`` indicates a request to call ``compute_vjp`` instead of ``compute_derivatives``. Note that 
```

if ``use_device_jacobian_product`` is ``True``, this takes precedence over calculating the full jacobian.

>>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
>>> processed_config = qml.device('default.qubit').preprocess(config)[1]
>>> processed_config.use_device_jacobian_product
True
>>> processed_config.use_device_gradient
True
>>> processed_config.grad_on_execution
True

Execution
---------

For documentation on the expected result type output, please refer to :ref:`ReturnTypeSpec`.

The device API allows individual devices to calculate results in whatever way makes sense for
the individual device. With this freedom over the implementation does come more responsibility
to handle each stage in the process. 

PennyLane does provide some helper functions to assist in executing
circuits. Any ``StateMeasurement`` has ``process_state`` and ``process_density_matrix`` methods for
classical post-processsing of a state vector or density matrix, and ``SampleMeasurement``'s implement

> both ``process_samples`` and ``process_counts``.  The ``pennylane.devices.qubit`` module also contains

Reviewer 2: ```suggestion
both ``process_samples`` and ``process_counts``. The ``pennylane.devices.qubit`` module also contains
```

functions that implement parts of a Python-based statevector simulation.

Suppose you are accessing hardware that can only return raw samples. Here, we use the ``mp.process_samples``

> methods to process the subsamples into the requested final result object.  Note that we need

Reviewer 2: ```suggestion
methods to process the subsamples into the requested final result object. Note that we need
```

inside ``execute`` instead of in ``preprocess``.

The :meth:`~.devices.Device.execute` can assume that device preprocessing has been run on the input
tapes, and has no obligation to re-validate the input or provide sensible error messages. In the below example,
we see `default.qubit` erroring out when unsupported operations and unsupported measurements are 

>>> op = qml.Permute([2,1,0], wires=(0,1,2))
>>> tape = qml.tape.QuantumScript([op], [qml.probs(wires=(0,1))])
>>> qml.device('default.qubit').execute(tape)
MatrixUndefinedError:
>>> tape = qml.tape.QuantumScript([], [qml.density_matrix(wires=0)], shots=50)
>>> qml.device('default.qubit').execute(tape)
AttributeError: 'DensityMatrixMP' object has no attribute 'process_samples'

Devices may define their own transforms following the description in the ``transforms`` module,
or can include in-built transforms such as:

.. currentmodule:: pennylane.devices.preprocess
.. toctree::
    :toctree: api

    decompose
    validate_observables
    validate_measurements
    validate_device_wires
    validate_multiprocessing_workers
    validate_adjoint_trainable_params
    no_sampling

.. currentmodule:: pennylane
.. toctree::

>     :toctree: api

Author: ```suggestion
```
